mlos_bench/mlos_bench/storage/sql_experiment.py

```diff
@@ -0,0 +1,178 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+"""
+Saving and restoring the benchmark data using SQLAlchemy.
+"""
+
+import logging
+import hashlib
+from datetime import datetime
+from typing import Optional, Tuple, List, Dict, Any
+
+from sqlalchemy import text, column
+
+from mlos_bench.tunables import TunableGroups
+from mlos_bench.storage.base_storage import Storage
+from mlos_bench.storage.sql_trial import Trial
+
+_LOG = logging.getLogger(__name__)
+
+# This is to allow passing all required values to the Experiment constructor
+# so that we don't have to expose the internals of the Storage class.
+# This constructor is not visible to the user, so it's OK to disable the warning.
+# pylint: disable=too-many-function-args,too-many-arguments
+
+
+class Experiment(Storage.Experiment):
+    """
+    Logic for retrieving and storing the results of a single experiment.
+    """
+
+    def __init__(self, engine, schema, tunables: TunableGroups,
+                 experiment_id: str, trial_id: int, description: str, opt_target: str):
+        super().__init__(tunables, experiment_id, trial_id, description, opt_target)
+        self._engine = engine
+        self._schema = schema
+
+    def __enter__(self):
+        super().__enter__()
+        with self._engine.begin() as conn:
+            # Get git info and the last trial ID for the experiment.
+            exp_info = conn.execute(
+                text("""
+                SELECT e.git_repo, e.git_commit, MAX(t.trial_id) AS trial_id
+                FROM experiment AS e
+                LEFT OUTER JOIN trial AS t ON (e.exp_id = t.exp_id)
+                WHERE e.exp_id = :exp_id
+                GROUP BY e.git_repo, e.git_commit
```

Comment:
I guess this is fine for now, but since we went through the effort of getting sql alchemy in use for schema definitions, does it also make sense to "just" use their ORM APIs as well now?

Reply:
Looks like we're already doing that elsewhere?

Reply:
I agree, let's stick to one standard. I'll push an update soon